lib/luaxtable.c

```diff
@@ -32,48 +14,190 @@
+		return NULL;
+
+	return (lunatik_object_t *)lunatik_toobject(L, -1);
 }
 
-static int luaxtable_match_check(const struct xt_mtchk_param *par)
+static inline int luaxtable_pushskb(lua_State *L, luaxtable_t *xtable, const struct sk_buff *skb, uint8_t opt)
 {
+	lunatik_object_t *data = luaxtable_getskb(L, xtable);
+	if (data == NULL) {
+		pr_err("could not find skb\n");
+		return -1;
+	}
+
+	luadata_reset(data, skb->data, skb_headlen(skb), opt);
 	return 0;
 }
 
-static void luaxtable_match_destroy(const struct xt_mtdtor_param *par)
+static inline int luaxtable_pushparams(lua_State *L, const struct xt_action_param *par, luaxtable_t *xtable, const struct sk_buff *skb, uint8_t opt)
 {
+	if (luaxtable_pushskb(L, xtable, skb, opt) == -1)
+		return -1;
+
+	lua_newtable(L);
+	lua_pushboolean(L, par->hotdrop);
+	lua_setfield(L, -2, "hotdrop");
+	lua_pushinteger(L, par->thoff);
+	lua_setfield(L, -2, "thoff");
+	lua_pushinteger(L, par->fragoff);
+	lua_setfield(L, -2, "fragoff");
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 10, 0))
+	lua_pushinteger(L, xt_hooknum(par));
+#else
+	lua_pushinteger(L, par->hooknum);
+#endif
+	lua_setfield(L, -2, "hooknum");
+	lua_pushvalue(L, -1); /* param table */
+	lua_insert(L, lua_gettop(L) - 2); /* stack: (...), param, skb, param */
+	return 0;
 }
 
-static unsigned int luaxtable_target(struct sk_buff *skb, const struct xt_action_param *par)
+static int luaxtable_domatch(lua_State *L, luaxtable_t *xtable, const struct sk_buff *skb, struct xt_action_param *par)
 {
-	return NF_ACCEPT;
+	int ret;
+	if (luaxtable_pushparams(L, par, xtable, skb, LUADATA_OPT_READONLY) == -1)
+		goto err;
+
+	if (luaxtable_invoke(L, xtable, "match", 2, 1) == -1)
+		goto err;
+
+	ret = lua_toboolean(L, -1);
+	lua_getfield(L, -2, "hotdrop");
+	par->hotdrop = lua_toboolean(L, -1);
+	return ret;
+err:
+	return luaxtable_hooks.match_fallback;
 }
 
-static int luaxtable_target_check(const struct xt_tgchk_param *par)
+static int luaxtable_dotarget(lua_State *L, luaxtable_t *xtable, struct sk_buff *skb, const struct xt_action_param *par)
 {
-	return 0;
+	int ret;
+	if (luaxtable_pushparams(L, par, xtable, skb, LUADATA_OPT_NONE) == -1)
+		goto err;
+
+	if (luaxtable_invoke(L, xtable, "target", 2, 1) == -1)
+		goto err;
+
+	ret = lua_tointeger(L, -1);
+	if (ret < 0 || ret > NF_MAX_VERDICT)
+		goto err;
+	return ret;
+err:
+	return luaxtable_hooks.target_fallback;
 }
 
-static void luaxtable_target_destroy(const struct xt_tgdtor_param *par)
+static int luaxtable_docheckentry(lua_State *L, luaxtable_t *xtable)
 {
+	if (luaxtable_invoke(L, xtable,"checkentry", 0, 1) == -1)
+		return -EINVAL;
+	return 0;
+}
+
+#define LUAXTABLE_HOOK_CB(T, U, V, hook, huk) 				\
+static T luaxtable_##hook(U skb, V par)					\
+{									\
+	int ret;							\
+	const luaxtable_info_t *info = (const luaxtable_info_t *)par->huk##info;	\
+	luaxtable_t *xtable = info->data;				\
+									\
+	lunatik_run(xtable->runtime, luaxtable_do##hook, ret, xtable, skb, par);	\
```

Comment:
you can actually pass `fallback` here, so you can avoid repeat code on `do{match,target}`